File: frontend/src/utils/measurementConverter.js
/**
 * Measurement conversion utility (frontend mirror of the backend MeasurementConverter).
 *
 * Volume units are normalised to millilitres and simplified to litres at >= 1000 ml.
 * Mass units are normalised to grams and simplified to kilograms at >= 1000 g.
 * Already-metric units pass through, and unrecognised / count-based units are left unchanged.
 */

const VOLUME_TO_ML = {
  tsp: 4.929, teaspoon: 4.929, teaspoons: 4.929,
  tbsp: 14.787, tablespoon: 14.787, tablespoons: 14.787,
  'fl oz': 29.574, 'fluid ounce': 29.574, 'fluid ounces': 29.574,
  cup: 236.588, cups: 236.588,
  pt: 473.176, pint: 473.176, pints: 473.176,
  qt: 946.353, quart: 946.353, quarts: 946.353,
  gal: 3785.41, gallon: 3785.41, gallons: 3785.41,
  l: 1000, liter: 1000, litre: 1000, liters: 1000, litres: 1000,
  ml: 1, milliliter: 1, millilitre: 1, milliliters: 1, millilitres: 1,
};

const MASS_TO_GRAMS = {
  oz: 28.3495, ounce: 28.3495, ounces: 28.3495,
  lb: 453.592, lbs: 453.592, pound: 453.592, pounds: 453.592,
  kg: 1000, kilogram: 1000, kilograms: 1000,
  g: 1, gram: 1, grams: 1,
};

function round(value, places) {
  const factor = 10 ** places;
  return Math.round(value * factor) / factor;
}

/**
 * Convert an (amount, unit) pair to its simplified metric representation.
 * @param {number} amount
 * @param {string} unit
 * @returns {{ amount: number, unit: string }}
 */
export function convertToMetric(amount, unit) {
  if (amount === null || amount === undefined || unit === null || unit === undefined) {
    return { amount, unit };
  }

  const normalized = String(unit).trim().toLowerCase();

  if (Object.prototype.hasOwnProperty.call(VOLUME_TO_ML, normalized)) {
    const milliliters = amount * VOLUME_TO_ML[normalized];
    if (milliliters >= 1000) {
      return { amount: round(milliliters / 1000, 2), unit: 'l' };
    }
    return { amount: round(milliliters, 0), unit: 'ml' };
  }

  if (Object.prototype.hasOwnProperty.call(MASS_TO_GRAMS, normalized)) {
    const grams = amount * MASS_TO_GRAMS[normalized];
    if (grams >= 1000) {
      return { amount: round(grams / 1000, 2), unit: 'kg' };
    }
    return { amount: round(grams, 0), unit: 'g' };
  }

  return { amount, unit };
}

/**
 * Format an amount/unit pair into a display string. All unit formatting in the UI
 * should route through this helper rather than being assembled inline in JSX.
 * @param {number} amount
 * @param {string} unit
 * @returns {string}
 */
export function formatMeasurement(amount, unit) {
  if (amount === null || amount === undefined) {
    return unit ? String(unit) : '';
  }
  const display = Number.isInteger(amount) ? amount : round(amount, 2);
  return unit ? `${display} ${unit}` : `${display}`;
}

/**
 * Convenience helper: convert to metric and immediately format for display.
 * @param {number} amount
 * @param {string} unit
 * @returns {string}
 */
export function formatAsMetric(amount, unit) {
  const converted = convertToMetric(amount, unit);
  return formatMeasurement(converted.amount, converted.unit);
}

File: backend/src/main/java/com/mealplanner/service/MeasurementConverter.java
package com.mealplanner.service;

import com.mealplanner.util.ConvertedMeasurement;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

/**
 * Converts ingredient measurements to the metric system.
 *
 * <p>Unambiguous liquid-volume units (fluid ounce, pint, quart, gallon, millilitre,
 * litre) are normalised to millilitres and simplified to litres when the result
 * reaches 1000 ml. Mass units are normalised to grams and simplified to kilograms
 * when the result reaches 1000 g. Units that are already metric pass through
 * unchanged.</p>
 *
 * <p>Household "scoop" units (teaspoon, tablespoon, cup) are intentionally NOT
 * converted: in everyday recipes they are used for solids and produce as well as
 * liquids (e.g. "2 cups spinach", "1 cup shredded cheese"), so turning them into
 * millilitres is misleading. They are therefore treated like count-based units
 * (e.g. "cloves", "pcs", "pinch") and returned untouched.</p>
 */
@Component
public class MeasurementConverter {

    private static final Map<String, Double> VOLUME_TO_ML = new HashMap<>();
    private static final Map<String, Double> MASS_TO_GRAMS = new HashMap<>();

    static {
        // Unambiguous liquid-volume variants -> millilitres.
        // Note: teaspoon/tablespoon/cup are deliberately excluded (see class Javadoc),
        // because they are commonly used to measure non-liquid ingredients too.
        registerAll(VOLUME_TO_ML, 29.574, "fl oz", "fluid ounce", "fluid ounces");
        registerAll(VOLUME_TO_ML, 473.176, "pt", "pint", "pints");
        registerAll(VOLUME_TO_ML, 946.353, "qt", "quart", "quarts");
        registerAll(VOLUME_TO_ML, 3785.41, "gal", "gallon", "gallons");
        registerAll(VOLUME_TO_ML, 1000.0, "l", "liter", "litre", "liters", "litres");
        registerAll(VOLUME_TO_ML, 1.0, "ml", "milliliter", "millilitre", "milliliters", "millilitres");

        // Mass variants -> grams
        registerAll(MASS_TO_GRAMS, 28.3495, "oz", "ounce", "ounces");
        registerAll(MASS_TO_GRAMS, 453.592, "lb", "lbs", "pound", "pounds");
        registerAll(MASS_TO_GRAMS, 1000.0, "kg", "kilogram", "kilograms");
        registerAll(MASS_TO_GRAMS, 1.0, "g", "gram", "grams");
    }

    private static void registerAll(Map<String, Double> target, double multiplier, String... variants) {
        for (String variant : variants) {
            target.put(variant, multiplier);
        }
    }

    /**
     * Converts an (amount, unit) pair to its simplified metric representation.
     *
     * @param amount the source amount; may be {@code null}
     * @param unit   the source unit; may be {@code null}
     * @return a {@link ConvertedMeasurement} with the converted amount and unit, or the
     * original values when the unit is not recognised as a volume or mass unit
     */
    public ConvertedMeasurement convert(Double amount, String unit) {
        if (amount == null || unit == null) {
            return new ConvertedMeasurement(amount, unit);
        }

        String normalized = unit.trim().toLowerCase();

        Double volumeMultiplier = VOLUME_TO_ML.get(normalized);
        if (volumeMultiplier != null) {
            double milliliters = amount * volumeMultiplier;
            if (milliliters >= 1000.0) {
                return new ConvertedMeasurement(round(milliliters / 1000.0, 2), "l");
            }
            return new ConvertedMeasurement(round(milliliters, 0), "ml");
        }

        Double massMultiplier = MASS_TO_GRAMS.get(normalized);
        if (massMultiplier != null) {
            double grams = amount * massMultiplier;
            if (grams >= 1000.0) {
                return new ConvertedMeasurement(round(grams / 1000.0, 2), "kg");
            }
            return new ConvertedMeasurement(round(grams, 0), "g");
        }

        // Unrecognised / count-based unit: leave unchanged.
        return new ConvertedMeasurement(amount, unit);
    }

    /**
     * Rounds a value to a fixed number of decimal places using half-up rounding.
     *
     * @param value  the value to round
     * @param places the number of decimal places (0 or more)
     * @return the rounded value
     */
    private double round(double value, int places) {
        double factor = Math.pow(10, places);
        return Math.round(value * factor) / factor;
    }
}
